### ISSUE TITLE
Add `ModuleRecoveryAlert` to authenticated dashboard

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#5256]
## Feature Description

Using `ModuleRecoveryAlert` implemented in #4823, the alert needs to be added to the authenticated dashboard to be visible to authenticated admins.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The `ModuleRecoveryAlert` component should be added to the main and entity dashboards, below any error notifications
* The component should only be rendered on the authenticated dashboards (not view-only)

## Implementation Brief

* * In `assets/js/components/Header.js` component:
    * Render `ModuleRecoveryAlert` under the `ErrorNotifications` component when the `dashboardSharing` feature flag is enabled and the dashboard is not `viewOnly`
    * `{ ! viewOnlyDashboard && dashboardSharingEnabled && <ModuleRecoveryAlert /> }`

### Test Coverage

* Add a story with the `ModuleRecoveryAlert` showing something (not `null`) in `assets/js/components/Header.stories.js`

### Site Kit

The new Sharing Settings plugin can be used to help show the Module Recovery Alert.

- Setup Site Kit and install the Sharing Settings plugin.
- Create a second admin user, log in to WordPress but don't sign in to Site Kit with the second user.
- As the second user, use the Sharing Settings plugin to change some settings for Pagespeed Insights and hit save.
  - This will have the effect of changing the owner ID for the PSI module. This can be verified with the following SQL query:
  - `SELECT option_value FROM wp_options WHERE option_name = 'googlesitekit_pagespeed-insights_settings';`
- View the Site Kit Dashboard as the first user. The Module Recovery Alert should be visible and pressing the Recover button should work (the PSI owner ID should be set to the first user's ID and the alert should disappear).
- In order to view the checkbox list variant of the Module Recovery Alert, enable the `ideaHubModule` feature flag and set up Idea Hub. Then follow the same steps as above, changing sharing settings as the second user for both PSI and Idea Hub. The alert should show up again for the first user with both modules listed.

### Storybook
 
* Ensure the Module Recovery Alert is visible in the new story **Components > Header > Plugin Header with Module Recovery Alert**.

## Changelog entry

* Add module recovery alert to the dashboard.

**Comment by eugene-manuilov** [comment]
IB ✔️

**Comment by techanvil** [comment]
Hi @aaemnnosttv, I've spotted that when I test this out on the Dashboard I get a `400` error from the `check-access` endpoint with the error _"Module access cannot be checked."_, for the `pagespeed-insights` module. The `ModuleRecoveryAlert` is then stuck in the loading state (see screenshot below). It should be easy to replicate if you just check out the branch, enable `dashboardSharing` and hit the Dashboard.

It looks like there's a bit of a disconnect between the list of modules returned by `get_recoverable_modules` vs the check that's made in `check-access`.

The logical execution path is as follows:

1. The result of `get_recoverable_modules` gets serialized to the page as `_googlesitekitDashboardSharingData.recoverableModules`.
2. The selector `getRecoverableModules` is bootstrapped with `_googlesitekitDashboardSharingData.recoverableModules`.
3. `ModuleRecoveryAlert` maps over `getRecoverableModules` calling `hasModuleAccess` for each module.
4. `hasModuleAccess` invokes the `check-access` endpoint for the given module.

In my testing I'm seeing the result of `get_recoverable_modules` as `['pagespeed-insights']`. The call to `check-access` fails because the PageSpeed Insights module does not implement `Module_With_Service_Entity`.

However there's no check to filter out modules that don't implement `Module_With_Service_Entity` when retrieving the list of recoverable modules, i.e. in `get_recoverable_modules`.

We could add a check to `is_module_recoverable` (this method is called via `get_recoverable_modules`) to return false if the specified module doesn't implement `Module_With_Service_Entity` to filter out those modules. However I am not entirely sure if this is the correct approach. Would appreciate your insight on this one.

![image](https://user-images.githubusercontent.com/18395600/170265342-083224bb-8517-4733-87b3-c2f83f7dee18.png)

**Comment by aaemnnosttv** [comment]
Thanks @techanvil – you're right, there is a bit of a problem there.

> However there's no check to filter out modules that don't implement `Module_With_Service_Entity` when retrieving the list of recoverable modules, i.e. in `get_recoverable_modules`.

The problem here is that PSI is intended to be recoverable, but does not have a service entity (i.e. an "owned" entity on its service, e.g. Analytics property).

The lack of a `Module_With_Service_Entity` interface what makes a shareable module a "shared ownership module" because there's nothing to share except the ownership itself (which determines who's credentials the request is made with). I think in this case, we need to tweak the logic in the `check-access` endpoint handler:

https://github.com/google/site-kit-wp/blob/762028639c26e106d56383874dee3b417a7f6693/includes/Core/Modules/Modules.php#L917-L931

- Move the check for the `Module_With_Service_Entity` interface after the check if the module is connected
- Then only return the `invalid_module` error if the module isn't shareable
- If it is shareable, and doesn't implement the interface, and is connected, then we can return `true` here

Ideally I think we would still make a basic request to the service to exercise that access but I think this is something we can refine in a follow up issue.

I realize this is a little out of scope here but I think the above change should be small enough to include here to unblock things. How does that sound to you?

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, that makes sense to me, I have gone ahead with the change as suggested.

Having resolved this issue I spotted another one - when pressing the **Recover** button, the call to `recover-module` is made successfully, but the banner stays visible, showing the CTA, even though the module(s) are recovered.

![image](https://user-images.githubusercontent.com/18395600/170489459-0aae9985-2063-40df-ad79-6e425e1ea1d9.png)

This is because the list of recoverable modules is not updated when the module is recovered. I've gone ahead and put a fix in place for this too, adding the following snippet to `recoverModule`:

https://github.com/google/site-kit-wp/blob/2c900857c5a9040e2e1a3e3f6035a4453b8476b1/assets/js/googlesitekit/modules/datastore/modules.js#L471-L483

It's a little out of keeping with the rest of `recoverModule` in that it updates state locally rather than refreshing it from the server. However we don't have an endpoint to retrieve the recoverable modules, so I figured this would be a more viable short term solution, at least, and we could consider updating it to refresh from the server in future.

Does that look OK to you?

**Comment by aaemnnosttv** [comment]
Hey @techanvil, thanks for raising this, your proposal sounds good to me as a solution for now 👍  Would you please open a follow-up issue to refresh the recoverable modules from the server as a future enhancement?

**Comment by techanvil** [comment]
> Hey @techanvil, thanks for raising this, your proposal sounds good to me as a solution for now +1 Would you please open a follow-up issue to refresh the recoverable modules from the server as a future enhancement?

Thanks @aaemnnosttv! I've created the followup issue https://github.com/google/site-kit-wp/issues/5287.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@techanvil Great QAB, thank you!  This does pass the QAB but have two observations:

1. The content layout seems off.  As you can see in the screenshot below, the first paragraph seems to be split into two columns, where as the second paragraph is one column.  At least thats what I can see when I inspect the CSS.

![image](https://user-images.githubusercontent.com/73545194/170661333-38bcbba4-4eb8-48be-80d6-397d01abe079.png)

2. The `Learn more` link URL is `/wp-admin/admin.php?page=googlesitekit-dashboard#` do we have a page URL in mind where the user should be redirected to? 

**Verified:**

- As the second user, when I changed settings for Pagespeed Insights, it has an effect of changing the owner ID for the PSI module. This was verified using the SQL query. 

- When viewing the Site Kit Dashboard as the first user, the Module Recovery Alert is visible and pressing the Recover button the PSI owner ID is set to the first user's ID and the alert disappears.

- Confirmed that the owner ID changed to the first admin user for the PSI module. This was verified using the SQL query. 

- The checkbox list appears on the Module Recovery Alert. Both Idea Hub and PSI modules are checked by default. When the recover button is clicked the PSI owner ID changes and the message disappears. 

-  The checkbox list appears on the Module Recovery Alert. Both Idea Hub and PSI modules are checked by default. When only one of the modules are selected, i.e. Idea Hub, and I click on the recover button, the message reappears with just the PSI mentioned in the alert.  The checklist disappears as expected. When I click on the recover button, the message disappears. 

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/170660025-da894e61-a0a2-46b2-acc9-66478c0903b6.png)

![image](https://user-images.githubusercontent.com/73545194/170661661-38d7878b-e11d-4fa4-8f14-ffe6e3216c4c.png)

![image](https://user-images.githubusercontent.com/73545194/170664990-83b4d79a-d67e-4dbb-a8d8-dd112072f00b.png)

</details>

**Comment by felixarntz** [comment]
### QA ❌ 

While testing the latest state of dashboard sharing, I noticed that the `ModuleRecoveryAlert` is currently being loaded on the splash screen in addition to the dashboard. It is not needed there, and it furthermore triggers a console error due to being loaded there since the necessary dashboard sharing data is (correctly) not being loaded on the splash screen.

I'll create a separate follow-up PR to address this.

**Comment by techanvil** [comment]
Nice one @felixarntz, I have merged your PR - good spot there!

**Comment by techanvil** [comment]
Just https://github.com/google/site-kit-wp/pull/5291 to go, I've addressed your comment on that one @aaemnnosttv.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil – back to you @wpdarren for another pass 👍

**Comment by wpdarren** [comment]
### QA Update: 

**Verified:**

- The content layout issue reported [above](https://github.com/google/site-kit-wp/issues/5256#issuecomment-1139417043) has been fixed. 
- The Learn more URL is now pointing to `https://sitekit.withgoogle.com/documentation/using-site-kit/dashboard-sharing/`

![image](https://user-images.githubusercontent.com/73545194/170948545-d491fbea-581c-47ed-bb73-cf5de20afe6b.png)

- I did notice the console error that @felixarntz mentioned but assumed (wrongly!) that it was because we are still building the feature. Can confirm that the message no longer appears. 
- As the second user, when I changed settings for Pagespeed Insights, it has an effect of changing the owner ID for the PSI module. This was verified using the SQL query. 
- When viewing the Site Kit Dashboard as the first user, the Module Recovery Alert is visible and pressing the Recover button the PSI owner ID is set to the first user's ID and the alert disappears.
- Confirmed that the owner ID changed to the first admin user for the PSI module. This was verified using the SQL query. 
- The checkbox list appears on the Module Recovery Alert. Both Idea Hub and PSI modules are checked by default. When the recover button is clicked the PSI owner ID changes and the message disappears. 
-  The checkbox list appears on the Module Recovery Alert. Both Idea Hub and PSI modules are checked by default. When only one of the modules is selected, i.e. Idea Hub, and I click on the recover button, the message reappears with just the PSI mentioned in the alert.  The checklist disappears. When I click on the recover button again, the message disappears. 

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/170660025-da894e61-a0a2-46b2-acc9-66478c0903b6.png)

![image](https://user-images.githubusercontent.com/73545194/170661661-38d7878b-e11d-4fa4-8f14-ffe6e3216c4c.png)

![image](https://user-images.githubusercontent.com/73545194/170664990-83b4d79a-d67e-4dbb-a8d8-dd112072f00b.png)

</details>
